File: packages/typescript/openlibx402-client/src/explicit-client.ts
/**
 * Explicit X402 Client
 *
 * Manual payment control - developer explicitly handles payment flow.
 */

import axios, { AxiosInstance, AxiosResponse, AxiosRequestConfig } from 'axios';
import { Keypair } from '@solana/web3.js';
import { URL } from 'url';
import {
  PaymentRequest,
  PaymentAuthorization,
  SolanaPaymentProcessor,
  PaymentExpiredError,
  InsufficientFundsError,
} from '@openlibx402/core';

export class X402Client {
  private walletKeypair: Keypair | null;
  private httpClient: AxiosInstance;
  private processor: SolanaPaymentProcessor;
  private _closed: boolean = false;
  private _allowLocal: boolean;

  /**
   * Explicit X402 client - developer controls payment flow
   *
   * Usage (Production):
   *   const client = new X402Client(walletKeypair);
   *
   *   // Check if payment required
   *   let response = await client.get("https://api.example.com/data");
   *
   *   if (client.paymentRequired(response)) {
   *     const paymentRequest = client.parsePaymentRequest(response);
   *
   *     // Make payment
   *     const authorization = await client.createPayment(paymentRequest);
   *
   *     // Retry with payment
   *     response = await client.get("https://api.example.com/data", { payment: authorization });
   *   }
   *
   *   // Always cleanup
   *   await client.close();
   *
   * Usage (Local Development):
   *   // Enable allowLocal for localhost URLs
   *   const client = new X402Client(walletKeypair, undefined, undefined, true);
   *
   *   let response = await client.get("http://localhost:3000/api/data");
   *
   *   if (client.paymentRequired(response)) {
   *     const paymentRequest = client.parsePaymentRequest(response);
   *     const authorization = await client.createPayment(paymentRequest);
   *     response = await client.get("http://localhost:3000/api/data", { payment: authorization });
   *   }
   *
   *   await client.close();
   *
   * Security Notes:
   *   - Always call close() when done to properly cleanup connections
   *   - Private keys are held in memory - ensure proper disposal
   *   - Only use URLs from trusted sources to prevent SSRF attacks
   *   - Default RPC URL is devnet - use mainnet URL for production
   *   - Set allowLocal=true for local development (localhost URLs)
   *   - NEVER use allowLocal=true in production deployments
   */
  constructor(
    walletKeypair: Keypair,
    rpcUrl?: string,
    httpClient?: AxiosInstance,
    allowLocal: boolean = false
  ) {
    this.walletKeypair = walletKeypair;
    this.httpClient = httpClient || axios.create({
      validateStatus: (status) => {
        // Accept 402 and 2xx status codes
        return status === 402 || (status >= 200 && status < 300);
      }
    });
    this.processor = new SolanaPaymentProcessor(
      rpcUrl || 'https://api.devnet.solana.com',
      walletKeypair
    );
    this._allowLocal = allowLocal;
  }

  /**
   * Close HTTP and RPC connections and cleanup sensitive data
   *
   * IMPORTANT: Always call this method when done to properly cleanup
   * connections and attempt to clear sensitive data from memory.
   */
  async close(): Promise<void> {
    if (this._closed) {
      return;
    }

    await this.processor.close();

    // Attempt to clear sensitive data (best-effort)
    this.walletKeypair = null;
    this._closed = true;
  }

  /**
   * Basic URL validation to prevent common SSRF attacks
   *
   * @throws {Error} If URL is invalid or potentially dangerous
   */
  private validateUrl(urlString: string): void {
    try {
      const url = new URL(urlString);

      // Require https or http scheme
      if (url.protocol !== 'http:' && url.protocol !== 'https:') {
        throw new Error(`Invalid URL scheme: ${url.protocol}. Only http/https allowed`);
      }

      // Skip localhost/private IP checks if allowLocal is enabled
      if (this._allowLocal) {
        return;
      }

      // Reject localhost and private IPs (basic check)
      const hostname = url.hostname.toLowerCase();

      // Check for localhost
      if (hostname === 'localhost' || hostname === '127.0.0.1' || hostname === '::1') {
        throw new Error(
          'Requests to localhost are not allowed. ' +
          'For local development, set allowLocal=true in constructor'
        );
      }

      // Check for private IP ranges (basic check)
      if (
        hostname.startsWith('10.') ||
        hostname.startsWith('192.168.') ||
        /^172\.(1[6-9]|2[0-9]|3[0-1])\./.test(hostname)
      ) {
        throw new Error(
          'Requests to private IP addresses are not allowed. ' +
          'For local development, set allowLocal=true in constructor'
        );
      }
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error(`Invalid URL: ${error}`);
    }
  }

  async get(
    url: string,
    options?: AxiosRequestConfig & { payment?: PaymentAuthorization }
  ): Promise<AxiosResponse> {
    this.validateUrl(url);
    const config = { ...options };
    if (options?.payment) {
      config.headers = {
        ...config.headers,
        'X-Payment-Authorization': options.payment.toHeaderValue(),
      };
      delete config.payment;
    }
    return this.httpClient.get(url, config);
  }

  async post(
    url: string,
    data?: any,
    options?: AxiosRequestConfig & { payment?: PaymentAuthorization }
  ): Promise<AxiosResponse> {
    this.validateUrl(url);
    const config = { ...options };
    if (options?.payment) {
      config.headers = {
        ...config.headers,
        'X-Payment-Authorization': options.payment.toHeaderValue(),
      };
      delete config.payment;
    }
    return this.httpClient.post(url, data, config);
  }

  async put(
    url: string,
    data?: any,
    options?: AxiosRequestConfig & { payment?: PaymentAuthorization }
  ): Promise<AxiosResponse> {
    this.validateUrl(url);
    const config = { ...options };
    if (options?.payment) {
      config.headers = {
        ...config.headers,
        'X-Payment-Authorization': options.payment.toHeaderValue(),
      };
      delete config.payment;
    }
    return this.httpClient.put(url, data, config);
  }

  async delete(
    url: string,
    options?: AxiosRequestConfig & { payment?: PaymentAuthorization }
  ): Promise<AxiosResponse> {
    this.validateUrl(url);
    const config = { ...options };
    if (options?.payment) {
      config.headers = {
        ...config.headers,
        'X-Payment-Authorization': options.payment.toHeaderValue(),
      };
      delete config.payment;
    }
    return this.httpClient.delete(url, config);
  }

  paymentRequired(response: AxiosResponse): boolean {
    return response.status === 402;
  }

  parsePaymentRequest(response: AxiosResponse): PaymentRequest {
    if (!this.paymentRequired(response)) {
      throw new Error('Response does not require payment (status !== 402)');
    }

    try {
      return PaymentRequest.fromDict(response.data);
    } catch (error) {
      throw new Error(`Failed to parse payment request: ${error}`);
    }
  }

  async createPayment(
    request: PaymentRequest,
    amount?: string
  ): Promise<PaymentAuthorization> {
    if (!this.walletKeypair) {
      throw new Error('Client has been closed');
    }

    // Validate request not expired (advisory check)
    if (request.isExpired()) {
      throw new PaymentExpiredError(request);
    }

    // Use provided amount or max required
    const payAmount = amount || request.maxAmountRequired;

    // Check sufficient balance
    const balance = await this.processor.getTokenBalance(
      this.walletKeypair.publicKey.toString(),
      request.assetAddress
    );

    // Convert to smallest unit (assuming 6 decimals) for precise integer comparison
    // This avoids floating-point precision issues with currency
    const decimals = 6;
    const balanceSmallestUnit = Math.floor(balance * Math.pow(10, decimals));
    const amountSmallestUnit = Math.floor(parseFloat(payAmount) * Math.pow(10, decimals));

    if (balanceSmallestUnit < amountSmallestUnit) {
      throw new InsufficientFundsError(payAmount, balance.toString());
    }

    // Create transaction
    const tx = await this.processor.createPaymentTransaction(
      request,
      payAmount,
      this.walletKeypair
    );

    // Sign and broadcast
    const txHash = await this.processor.signAndSendTransaction(
      tx,
      this.walletKeypair
    );

    // Create authorization
    // Note: In Solana, the transaction signature IS the transaction hash
    return new PaymentAuthorization({
      payment_id: request.paymentId,
      actual_amount: payAmount,
      payment_address: request.paymentAddress,
      asset_address: request.assetAddress,
      network: request.network,
      timestamp: new Date(),
      signature: txHash, // Solana transaction signature
      public_key: this.walletKeypair.publicKey.toString(),
      transaction_hash: txHash,
    });
  }

  async request(
    method: string,
    url: string,
    options?: AxiosRequestConfig & { payment?: PaymentAuthorization }
  ): Promise<AxiosResponse> {
    const methodUpper = method.toUpperCase();

    switch (methodUpper) {
      case 'GET':
        return this.get(url, options);
      case 'POST':
        return this.post(url, undefined, options);
      case 'PUT':
        return this.put(url, undefined, options);
      case 'DELETE':
        return this.delete(url, options);
      default:
        throw new Error(`Unsupported HTTP method: ${method}`);
    }
  }
}

File: packages/go/openlibx402-client/explicit_client.go
// Package client provides HTTP client functionality for the X402 payment protocol.
//
// It includes both explicit and automatic payment handling for AI agents and applications
// that need to make payments to access paid APIs.
package client

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"net/url"
	"strings"

	"github.com/gagliardetto/solana-go"
	"github.com/openlibx402/go/openlibx402-core"
)

// X402Client provides explicit control over the X402 payment flow.
//
// With explicit mode, the developer manually checks for 402 responses,
// creates payments, and retries requests with payment authorization.
type X402Client struct {
	walletKeypair *solana.PrivateKey
	httpClient    *http.Client
	processor     *core.SolanaPaymentProcessor
	allowLocal    bool
	closed        bool
}

// NewX402Client creates a new explicit X402 client.
//
// Parameters:
//   - walletKeypair: Solana wallet keypair for signing transactions
//   - rpcURL: Solana RPC endpoint URL (optional, defaults to devnet)
//   - httpClient: Custom HTTP client (optional)
//   - allowLocal: Allow requests to localhost/private IPs (for development only)
//
// Usage (Production):
//
//	client := NewX402Client(walletKeypair, "", nil, false)
//	defer client.Close()
//
//	resp, err := client.Get(ctx, "https://api.example.com/data", nil)
//	if err != nil {
//	    log.Fatal(err)
//	}
//
//	if client.PaymentRequired(resp) {
//	    paymentReq, err := client.ParsePaymentRequest(resp)
//	    if err != nil {
//	        log.Fatal(err)
//	    }
//
//	    auth, err := client.CreatePayment(ctx, paymentReq, "")
//	    if err != nil {
//	        log.Fatal(err)
//	    }
//
//	    resp, err = client.Get(ctx, "https://api.example.com/data", auth)
//	    if err != nil {
//	        log.Fatal(err)
//	    }
//	}
//
// Usage (Local Development):
//
//	client := NewX402Client(walletKeypair, "", nil, true)
//	defer client.Close()
//
//	resp, err := client.Get(ctx, "http://localhost:8080/api/data", nil)
//	// ... handle payment flow same as above
func NewX402Client(
	walletKeypair solana.PrivateKey,
	rpcURL string,
	httpClient *http.Client,
	allowLocal bool,
) *X402Client {
	if rpcURL == "" {
		rpcURL = "https://api.devnet.solana.com"
	}

	if httpClient == nil {
		httpClient = &http.Client{}
	}

	processor := core.NewSolanaPaymentProcessor(rpcURL, &walletKeypair)

	return &X402Client{
		walletKeypair: &walletKeypair,
		httpClient:    httpClient,
		processor:     processor,
		allowLocal:    allowLocal,
		closed:        false,
	}
}

// Close closes the client and cleans up resources.
//
// IMPORTANT: Always call this method when done to properly cleanup
// connections and attempt to clear sensitive data from memory.
func (c *X402Client) Close() error {
	if c.closed {
		return nil
	}

	err := c.processor.Close()
	c.walletKeypair = nil
	c.closed = true
	return err
}

// validateURL performs basic URL validation to prevent SSRF attacks.
func (c *X402Client) validateURL(urlStr string) error {
	parsedURL, err := url.Parse(urlStr)
	if err != nil {
		return fmt.Errorf("invalid URL: %w", err)
	}

	// Require http or https scheme
	if parsedURL.Scheme != "http" && parsedURL.Scheme != "https" {
		return fmt.Errorf("invalid URL scheme: %s. Only http/https allowed", parsedURL.Scheme)
	}

	// Skip localhost/private IP checks if allowLocal is enabled
	if c.allowLocal {
		return nil
	}

	hostname := strings.ToLower(parsedURL.Hostname())

	// Check for localhost
	if hostname == "localhost" || hostname == "127.0.0.1" || hostname == "::1" {
		return fmt.Errorf("requests to localhost are not allowed. For local development, set allowLocal=true")
	}

	// Check for private IP ranges (basic check)
	if strings.HasPrefix(hostname, "10.") ||
		strings.HasPrefix(hostname, "192.168.") ||
		strings.HasPrefix(hostname, "172.16.") ||
		strings.HasPrefix(hostname, "172.17.") ||
		strings.HasPrefix(hostname, "172.18.") ||
		strings.HasPrefix(hostname, "172.19.") ||
		strings.HasPrefix(hostname, "172.20.") ||
		strings.HasPrefix(hostname, "172.21.") ||
		strings.HasPrefix(hostname, "172.22.") ||
		strings.HasPrefix(hostname, "172.23.") ||
		strings.HasPrefix(hostname, "172.24.") ||
		strings.HasPrefix(hostname, "172.25.") ||
		strings.HasPrefix(hostname, "172.26.") ||
		strings.HasPrefix(hostname, "172.27.") ||
		strings.HasPrefix(hostname, "172.28.") ||
		strings.HasPrefix(hostname, "172.29.") ||
		strings.HasPrefix(hostname, "172.30.") ||
		strings.HasPrefix(hostname, "172.31.") {
		return fmt.Errorf("requests to private IP addresses are not allowed. For local development, set allowLocal=true")
	}

	return nil
}

// Do executes an HTTP request with optional payment authorization.
func (c *X402Client) Do(ctx context.Context, req *http.Request, payment *core.PaymentAuthorization) (*http.Response, error) {
	if c.closed {
		return nil, fmt.Errorf("client has been closed")
	}

	if err := c.validateURL(req.URL.String()); err != nil {
		return nil, err
	}

	// Add payment authorization header if provided
	if payment != nil {
		headerValue, err := payment.ToHeaderValue()
		if err != nil {
			return nil, fmt.Errorf("failed to encode payment authorization: %w", err)
		}
		req.Header.Set("X-Payment-Authorization", headerValue)
	}

	// Execute request
	return c.httpClient.Do(req.WithContext(ctx))
}

// Get executes a GET request.
func (c *X402Client) Get(ctx context.Context, url string, payment *core.PaymentAuthorization) (*http.Response, error) {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}
	return c.Do(ctx, req, payment)
}

// Post executes a POST request.
func (c *X402Client) Post(ctx context.Context, url string, body []byte, payment *core.PaymentAuthorization) (*http.Response, error) {
	req, err := http.NewRequest("POST", url, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	return c.Do(ctx, req, payment)
}

// Put executes a PUT request.
func (c *X402Client) Put(ctx context.Context, url string, body []byte, payment *core.PaymentAuthorization) (*http.Response, error) {
	req, err := http.NewRequest("PUT", url, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	return c.Do(ctx, req, payment)
}

// Delete executes a DELETE request.
func (c *X402Client) Delete(ctx context.Context, url string, payment *core.PaymentAuthorization) (*http.Response, error) {
	req, err := http.NewRequest("DELETE", url, nil)
	if err != nil {
		return nil, err
	}
	return c.Do(ctx, req, payment)
}

// PaymentRequired checks if a response indicates that payment is required (status 402).
func (c *X402Client) PaymentRequired(resp *http.Response) bool {
	return resp.StatusCode == http.StatusPaymentRequired
}

// ParsePaymentRequest parses a PaymentRequest from a 402 response.
func (c *X402Client) ParsePaymentRequest(resp *http.Response) (*core.PaymentRequest, error) {
	if !c.PaymentRequired(resp) {
		return nil, fmt.Errorf("response does not require payment (status != 402)")
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}
	defer resp.Body.Close()

	var paymentReq core.PaymentRequest
	if err := json.Unmarshal(body, &paymentReq); err != nil {
		return nil, core.NewInvalidPaymentRequestError("failed to parse payment request: " + err.Error())
	}

	return &paymentReq, nil
}

// CreatePayment creates and broadcasts a payment transaction, returning a PaymentAuthorization.
//
// Parameters:
//   - ctx: Context for cancellation
//   - request: The payment request from the 402 response
//   - amount: Optional custom amount (uses max_amount_required if empty)
//
// Returns:
//   - PaymentAuthorization to include in retry request
func (c *X402Client) CreatePayment(
	ctx context.Context,
	request *core.PaymentRequest,
	amount string,
) (*core.PaymentAuthorization, error) {
	if c.closed {
		return nil, fmt.Errorf("client has been closed")
	}

	if c.walletKeypair == nil {
		return nil, fmt.Errorf("client has been closed")
	}

	// Validate request not expired
	if request.IsExpired() {
		return nil, core.NewPaymentExpiredError(request, "")
	}

	// Use provided amount or max required
	payAmount := amount
	if payAmount == "" {
		payAmount = request.MaxAmountRequired
	}

	// Check sufficient balance
	balance, err := c.processor.GetTokenBalance(
		ctx,
		c.walletKeypair.PublicKey().String(),
		request.AssetAddress,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to get token balance: %w", err)
	}

	// Convert to smallest unit (assuming 6 decimals) for precise comparison
	decimals := 6
	balanceSmallestUnit := uint64(math.Floor(balance * math.Pow(10, float64(decimals))))

	payAmountFloat := 0.0
	_, err = fmt.Sscanf(payAmount, "%f", &payAmountFloat)
	if err != nil {
		return nil, fmt.Errorf("invalid amount format: %w", err)
	}
	amountSmallestUnit := uint64(math.Floor(payAmountFloat * math.Pow(10, float64(decimals))))

	if balanceSmallestUnit < amountSmallestUnit {
		return nil, core.NewInsufficientFundsError(payAmount, fmt.Sprintf("%.6f", balance))
	}

	// Create transaction
	tx, err := c.processor.CreatePaymentTransaction(ctx, request, payAmount, *c.walletKeypair)
	if err != nil {
		return nil, err
	}

	// Sign and broadcast
	txHash, err := c.processor.SignAndSendTransaction(ctx, tx, *c.walletKeypair)
	if err != nil {
		return nil, err
	}

	// Create authorization
	return &core.PaymentAuthorization{
		PaymentID:       request.PaymentID,
		ActualAmount:    payAmount,
		PaymentAddress:  request.PaymentAddress,
		AssetAddress:    request.AssetAddress,
		Network:         request.Network,
		Timestamp:       request.ExpiresAt, // Use current time in production
		Signature:       txHash,
		PublicKey:       c.walletKeypair.PublicKey().String(),
		TransactionHash: txHash,
	}, nil
}
